### PR TITLE
Text sounds

### PR DESCRIPTION
**Pull request by ewenw**
Hi, 

I moved the text and sound functions from WaitingRoom widget into all the widgets. 

-Ewen

### CODE REVIEW

File: lib/Widgets.js [hunk 1]
```diff
@@ -240,6 +240,8 @@
             WidgetPrototype.className : options.className;
         widget.context = 'undefined' === typeof options.context ?
             WidgetPrototype.context : options.context;
+        widget.sounds = 'undefined' === typeof options.sounds ?
+            WidgetPrototype.sounds : options.sounds;
```
Comment (shakty) [inline lib/Widgets.js]: Is there a reason why you didn't do the same for .texts?

File: widgets/WaitingRoom.js [hunk 2]
```diff
@@ -652,337 +652,4 @@
...
-            that[collection][name] = value;
-        }
-        else {
-            throw new TypeError(method + ': value for item "' + name +
-                                '" must be string, function or false. ' +
-                                'Found: ' + value);
-        }
-    }
-
 })(node);
```
Comment (shakty) [inline widgets/WaitingRoom.js]: You should remove also the property from the constructor: .texts and .sounds are automatically added by the Widgets.js file now.

Comment (ewenw) [reply widgets/WaitingRoom.js]: Do you mean the WaitingRoom.text = {...} and WaitingRoom.sounds = {...} properties should be removed?

Comment (shakty) [reply widgets/WaitingRoom.js]: I mean this:

        /**
         * ### WaitingRoom.texts
         *
         * Contains all the texts displayed to the players
         *
         * @see WaitingRoom.setText
         * @see WaitingRoom.getText
         * @see WaitingRoom.setTexts
         * @see WaitingRoom.getTexts
         */
        this.texts = {};

        /**
         * ### WaitingRoom.sounds
         *
         * List of custom sounds to play to the players
         *
         * @see WaitingRoom.setSound
         * @see WaitingRoom.getSound
         * @see WaitingRoom.setSounds
         * @see WaitingRoom.getSounds
         */
         this.sounds = {};

Make sure the @see comments are properly copied in Widget.js

File: widgets/EmailForm.js [hunk 3]
```diff
@@ -86,8 +53,8 @@
         }
         else {
             throw new TypeError('EmailForm constructor: options.onsubmit ' +
-                                'must be string or object. Found: ' +
-                                options.onsubmit);
+                'must be string or object. Found: ' +
+                options.onsubmit);
```
Comment (shakty) [inline widgets/EmailForm.js]: Align text.

File: widgets/EmailForm.js [hunk 4]
```diff
@@ -289,7 +256,7 @@
         email = getEmail.call(this);
 
         if (opts.verify !== false) res = this.verifyInput(opts.markAttempt,
-                                                          opts.updateUI);
+            opts.updateUI);
```
Comment (shakty) [inline widgets/EmailForm.js]: Align text.

File: widgets/EmailForm.js [hunk 5]
```diff
@@ -353,7 +320,7 @@
         if (!this.inputElement) return;
         if (border && 'string' !== typeof border) {
             throw new TypeError('EmailForm.highlight: border must be ' +
-                                'string or undefined. Found: ' + border);
+                'string or undefined. Found: ' + border);
```
Comment (shakty) [inline widgets/EmailForm.js]: Align text.

File: widgets/EndScreen.js [hunk 6]
```diff
@@ -179,7 +145,7 @@
             this.totalWinCurrency = 'USD';
         }
         else if ('string' === typeof options.totalWinCurrency &&
-                 options.totalWinCurrency.trim() !== '') {
+            options.totalWinCurrency.trim() !== '') {
```
Comment (shakty) [inline widgets/EndScreen.js]: Align text.

File: widgets/EndScreen.js [hunk 7]
```diff
@@ -245,18 +215,20 @@
...
         endScreenElement.appendChild(messageElement);
 
         if (this.showTotalWin) {
             totalWinElement = document.createElement('div');
 
             totalWinParaElement = document.createElement('p');
-            totalWinParaElement.innerHTML = '<strong>Your total win:</strong>';
+            totalWinParaElement.innerHTML = '<strong>' +
+                this.getText('total_win') +
+                '</strong>';
```
Comment (shakty) [inline widgets/EndScreen.js]: align text.

File: widgets/EndScreen.js [hunk 8]
```diff
@@ -273,11 +245,13 @@
             exitCodeElement = document.createElement('div');
 
             exitCodeParaElement = document.createElement('p');
-            exitCodeParaElement.innerHTML = '<strong>Your exit code:</strong>';
+            exitCodeParaElement.innerHTML = '<strong>' +
+                this.getText('exit_code') +
```
Comment (shakty) [inline widgets/EndScreen.js]: align, maybe can make it in two lines

File: widgets/EndScreen.js [hunk 9]
```diff
@@ -273,11 +245,13 @@
...
+                '</strong>';
 
             exitCodeInputElement = document.createElement('input');
             exitCodeInputElement.className = 'endscreen-exit-code ' +
-                                             'form-control';
+                'form-control';
```
Comment (shakty) [inline widgets/EndScreen.js]: align

File: widgets/EndScreen.js [hunk 10]
```diff
@@ -328,7 +302,7 @@
             else if (data.partials) {
                 if (!J.isArray(data.partials)) {
                     node.err('EndScreen error, invalid partials win: ' +
-                             data.partials);
+                        data.partials);
```
Comment (shakty) [inline widgets/EndScreen.js]: align

File: lib/Widget.js [hunk 11]
```diff
@@ -288,9 +288,9 @@
                 }
                 else if ('object' !== typeof options) {
                     throw new TypeError('Widget.setTitle: options must ' +
-                                        'be object or undefined. Found: ' +
-                                        options);                    
-                }                
+                        'be object or undefined. Found: ' +
```
Comment (shakty) [inline lib/Widget.js]: align

File: lib/Widget.js [hunk 12]
```diff
@@ -308,8 +308,8 @@
             }
             else {
                 throw new TypeError(J.funcName(this.constructor) +
-                                    '.setTitle: title must be string, ' +
-                                    'HTML element or falsy. Found: ' + title);
+                    '.setTitle: title must be string, ' +
```
Comment (shakty) [inline lib/Widget.js]: align

File: lib/Widget.js [hunk 13]
```diff
@@ -349,8 +349,8 @@
                 }
                 else if ('object' !== typeof options) {
                     throw new TypeError('Widget.setFooter: options must ' +
-                                        'be object or undefined. Found: ' +
-                                        options);                    
+                        'be object or undefined. Found: ' +
```
Comment (shakty) [inline lib/Widget.js]: align

File: lib/Widget.js [hunk 14]
```diff
@@ -384,59 +384,403 @@
     Widget.prototype.setContext = function(context) {
         if ('string' !== typeof context) {
             throw new TypeError(J.funcName(this.constructor) + '.setContext: ' +
-                                'context must be string. Found: ' + context);
+                'context must be string. Found: ' + context);
```
Comment (shakty) [inline lib/Widget.js]: align

File: lib/Widget.js [hunk 15]
```diff
@@ -366,8 +366,8 @@
             }
             else {
                 throw new TypeError(J.funcName(this.constructor) +
-                                    '.setFooter: footer must be string, ' +
-                                    'HTML element or falsy. Found: ' + title);
+                    '.setFooter: footer must be string, ' +
```
Comment (shakty) [inline lib/Widget.js]: align

File: lib/Widget.js [hunk 16]
```diff
@@ -384,59 +384,403 @@
...
+     * @see Widget.panelDiv
+     * @see Widget.bodyDiv
+     */
+    Widget.prototype.addFrame = function(context) {
+        if ('undefined' === typeof context) {
+            context = 'default';
+        }
+        else if ('string' !== typeof context || context.trim() === '') {
+            throw new TypeError(J.funcName(this.constructor) +
+                '.addFrame: context must be a non-empty ' +
```
Comment (shakty) [inline lib/Widget.js]: align

File: lib/Widget.js [hunk 17]
```diff
@@ -384,59 +384,403 @@
...
+     * @return {object} Selected sounds (paths)
+     *
+     * @see Widget.sounds
+     * @see Widget.setSound
+     * @see Widget.getSound
+     * @see Widget.setSounds
+     */
+    Widget.prototype.getSounds = function(keys, param) {
+        return strGetterMulti(this, 'sounds', 'getSound',
+            J.funcName(this.constructor)
```
Comment (shakty) [inline lib/Widget.js]: align

File: lib/Widget.js [hunk 18]
```diff
@@ -384,59 +384,403 @@
...
+     * @param {object} param Optional. Object containing parameters to pass
+     *   to the sounds functions (if any)
+     *
+     * @return {object} All current sounds
+     *
+     * @see Widget.getSound
+     */
+    Widget.prototype.getAllSounds = function(param) {
+        return strGetterMulti(this, 'sounds', 'getSound',
+            J.funcName(this.constructor)
```
Comment (shakty) [inline lib/Widget.js]: PLUS must be at the end of previous line

**Pull request closed without merging**